File: 2015/22/rust/src/main.rs
use enum_derive_list::AllVariants;
use regex::Regex;

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";
const PARSER_REGEX: &'static str = r"Hit Points: (\d+).*Damage: (\d+)";
const PLAYER_HP: isize = 50;
const PLAYER_MANA: usize = 500;
const MAGIC_MISSILE: Magic  = Magic { cost:  53, duration: 0, effect: 4 };
const MAGIC_DRAIN: Magic    = Magic { cost:  73, duration: 0, effect: 2 };
const MAGIC_SHIELD: Magic   = Magic { cost: 113, duration: 6, effect: 7 };
const MAGIC_POISON: Magic   = Magic { cost: 173, duration: 6, effect: 3 };
const MAGIC_RECHARGE: Magic = Magic { cost: 229, duration: 5, effect: 101 };

// --- Enums ---
#[derive(Debug, PartialEq)]
enum PlayResult { Win, Lose, Continue }

#[derive(AllVariants, Debug, PartialEq)]
enum Spell { Missile, Drain, Shield, Poison, Recharge }

impl Spell {
    fn magic(&self) -> &Magic {
        match self {
            Spell::Missile  => &MAGIC_MISSILE,
            Spell::Drain    => &MAGIC_DRAIN,
            Spell::Shield   => &MAGIC_SHIELD,
            Spell::Poison   => &MAGIC_POISON,
            Spell::Recharge => &MAGIC_RECHARGE
        }
    }
}

// --- Structs ---
#[derive(Debug)]
struct Magic {
    cost: usize,
    duration: usize,
    effect: usize
}

#[derive(Debug, Clone)]
struct State {
    p_hp: isize,
    p_armor: usize,
    p_mana: usize,
    shield: usize,
    poison: usize,
    recharge: usize,
    b_hp: isize,
    used_mana: usize
}

impl State {
    fn new(b_hp: isize) -> Self {
        State { p_hp: PLAYER_HP, p_armor: 0, p_mana: PLAYER_MANA, shield: 0, poison: 0, recharge: 0, b_hp, used_mana: 0 }
    }
}

// --- Checks if the player can cast a spell ---
fn can_cast(state: &State, spell: &Spell) -> bool {
    if state.p_mana < spell.magic().cost { return false; }
    match spell {
        Spell::Shield   => state.shield == 0,
        Spell::Poison   => state.poison == 0,
        Spell::Recharge => state.recharge == 0,
        Spell::Missile | Spell::Drain => true
    }
}

// --- Apply the current active effects ---
fn apply_effects(state: &mut State) {
    if state.shield > 0 {
        state.shield -= 1;
        state.p_armor = if state.shield > 0 { Spell::Shield.magic().effect } else { 0 };
    }
    if state.poison >  0 {
        state.poison -= 1;
        state.b_hp -= Spell::Poison.magic().effect as isize;
    }
    if state.recharge > 0 {
        state.recharge -= 1;
        state.p_mana += Spell::Recharge.magic().effect;
    }
}

// --- Cast a spell and simulate boss turn ---
fn cast(state: &mut State, spell: &Spell, boss_damage: usize) -> PlayResult {
    // --- Player turn ---
    let magic = spell.magic();
    state.p_mana -= magic.cost;
    state.used_mana += magic.cost;
    match spell {
        Spell::Missile => state.b_hp -= magic.effect as isize,
        Spell::Drain => {
            state.b_hp -= magic.effect as isize;
            state.p_hp += magic.effect as isize;
        },
        Spell::Shield   => state.shield = magic.duration,
        Spell::Poison   => state.poison = magic.duration,
        Spell::Recharge => state.recharge = magic.duration
    }

    // --- Boss turn ---
    apply_effects(state);
    if state.b_hp <= 0 { return PlayResult::Win; }

    let hit = std::cmp::max(boss_damage as isize - state.p_armor as isize, 1);
    state.p_hp -= hit;
    if state.p_hp <= 0 { return PlayResult::Lose; }

    PlayResult::Continue
}

// --- Play the game ---
fn play(b_hp: isize, b_damage: usize, hardmode: bool) -> usize {
    let mut queue: Vec<State> = Vec::new();
    let mut min = std::usize::MAX;

    // --- Search winning games ---
    queue.push(State::new(b_hp));
    while queue.len() > 0 {
        let mut state = queue.pop().unwrap();

        // --- Ignore worst paths ---
        if state.used_mana >= min { continue; }

        // --- Hard mode ---
        if hardmode {
            state.p_hp -= 1;
            if state.p_hp <= 0 { continue; }
        }

        // --- Apply player turn effects ---
        apply_effects(&mut state);
        if state.b_hp <= 0 {
            min = std::cmp::min(min, state.used_mana);
            continue;
        }

        // --- Try every spell except Missile ---
        for spell in Spell::all_variants() {
            if spell == &Spell::Missile { continue; }
            if can_cast(&state, spell) {
                let mut clone = state.clone();
                let result = cast(&mut clone, spell, b_damage);
                match result {
                    PlayResult::Win      => min = std::cmp::min(min, clone.used_mana),
                    PlayResult::Lose     => (),
                    PlayResult::Continue => queue.push(clone)
                }
            }
        }

        // --- Try cast Missile ---
        if can_cast(&state, &Spell::Missile) {
            match cast(&mut state, &Spell::Missile, b_damage) {
                PlayResult::Win      => min = std::cmp::min(min, state.used_mana),
                PlayResult::Lose     => (),
                PlayResult::Continue => queue.push(state)
            }
        }
    }
    min
}

// --- Main function ---
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --- Read and parse the input file ---
    let regex = Regex::new(PARSER_REGEX)?;
    let text = std::fs::read_to_string(INPUT_FILE)?.replace("\n", " ");
    let cap = regex.captures(&text).unwrap();
    let health: isize = cap[1].parse()?;
    let damage: usize = cap[2].parse()?;

    let min = play(health, damage, false);
    println!("1. Least amount of mana: {}", min);

    let min = play(health, damage, true);
    println!("2. Least amount of mana: {}", min);

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cast_1() {
        let damage = 8;
        let mut state = State::new(13);
        state.p_hp = 10;
        state.p_mana = 250;
        apply_effects(&mut state);
        let result = cast(&mut state, &Spell::Poison, damage);
        assert_eq!(result, PlayResult::Continue);
        assert_state(&state, 2, 0, 77, 0, 5, 0, 10, 173);
        apply_effects(&mut state);
        let result = cast(&mut state, &Spell::Missile, damage);
        assert_eq!(result, PlayResult::Win);
        assert_state(&state, 2, 0, 24, 0, 3, 0, 0, 173 + 53);
    }

    #[test]
    fn test_cast_2() {
        let damage = 8;
        let mut state = State::new(14);
        state.p_hp = 10;
        state.p_mana = 250;
        apply_effects(&mut state);
        let result = cast(&mut state, &Spell::Recharge, damage);
        assert_eq!(result, PlayResult::Continue);
        assert_state(&state, 2, 0, 122, 0, 0, 4, 14, 229);
        apply_effects(&mut state);
        let result = cast(&mut state, &Spell::Shield, damage);
        assert_eq!(result, PlayResult::Continue);
        assert_state(&state, 1, 7, 211, 5, 0, 2, 14, 342);
        apply_effects(&mut state);
        let result = cast(&mut state, &Spell::Drain, damage);
        assert_eq!(result, PlayResult::Continue);
        assert_state(&state, 2, 7, 340, 3, 0, 0, 12, 415);
        apply_effects(&mut state);
        let result = cast(&mut state, &Spell::Poison, damage);
        assert_eq!(result, PlayResult::Continue);
        assert_state(&state, 1, 7, 167, 1, 5, 0, 9, 588);
        apply_effects(&mut state);
        let result = cast(&mut state, &Spell::Missile, damage);
        assert_eq!(result, PlayResult::Win);
        assert_state(&state, 1, 0, 114, 0, 3, 0, -1, 641);
    }

    fn assert_state(state: &State, p_hp: isize, p_armor: usize, p_mana: usize, shield: usize, poison: usize, recharge: usize, b_hp: isize, used_mana: usize) {
        assert_eq!(state.p_hp, p_hp);
        assert_eq!(state.p_armor, p_armor);
        assert_eq!(state.p_mana, p_mana);
        assert_eq!(state.shield, shield);
        assert_eq!(state.poison, poison);
        assert_eq!(state.recharge, recharge);
        assert_eq!(state.b_hp, b_hp);
        assert_eq!(state.used_mana, used_mana);
    }
}

File: 2015/22/python/main.py
from enum import Enum
import copy
import re

# --- Constants --
INPUT_FILE = "../input/input.txt"
REGEX = r"Hit Points: (\d+).*Damage: (\d+)"
PLAYER_HP = 50
PLAYER_MANA = 500

# --- Enumerations ---
Result = Enum('Result', ['WIN', 'LOSE', 'CONTINUE'])

# --- Classes ---
class Magic:
    def __init__(self, cost, duration, effect):
        self.cost = cost
        self.duration = duration
        self.effect = effect

class State:
    def __init__(self, player_hp, player_mana, boss_hp):
        self.player_hp = player_hp
        self.player_mana = player_mana
        self.player_armor = 0
        self.shield = 0
        self.poison = 0
        self.recharge = 0
        self.boss_hp = boss_hp
        self.used_mana = 0

    def __str__(self):
        return f"{self.player_hp}, {self.player_mana}, {self.player_armor}, {self.shield}, {self.poison}, {self.recharge}, {self.boss_hp}, {self.used_mana}"

# --- Magic Constants ---
MAGIC_MISSILE  = Magic( 53, 0,   4);
MAGIC_DRAIN    = Magic( 73, 0,   2);
MAGIC_SHIELD   = Magic(113, 6,   7);
MAGIC_POISON   = Magic(173, 6,   3);
MAGIC_RECHARGE = Magic(229, 5, 101);

# --- Functions ---
def apply_effects(state):
    if state.shield > 0:
        state.shield -= 1
        state.player_armor = MAGIC_SHIELD.effect if state.shield > 0 else 0
    if state.poison > 0:
        state.poison -= 1
        state.boss_hp -= MAGIC_POISON.effect
    if state.recharge > 0:
        state.recharge -= 1
        state.player_mana += MAGIC_RECHARGE.effect

def can_cast(state, magic):
    if state.player_mana < magic.cost: return False
    if magic == MAGIC_SHIELD: return state.shield == 0
    if magic == MAGIC_POISON: return state.poison == 0
    if magic == MAGIC_RECHARGE: return state.recharge == 0
    return True

def cast(state, magic, boss_damage):
    # --- Players turn ---
    state.player_mana -= magic.cost
    state.used_mana += magic.cost
    if magic == MAGIC_MISSILE:
        state.boss_hp -= magic.effect
    elif magic == MAGIC_DRAIN:
        state.boss_hp -= magic.effect
        state.player_hp += magic.effect
    elif magic == MAGIC_SHIELD:
        state.shield = magic.duration
    elif magic == MAGIC_POISON:
        state.poison = magic.duration
    elif magic == MAGIC_RECHARGE:
        state.recharge = magic.duration
    else:
        raise Exception("Unsupported magic")

    # --- Boss turn ---
    apply_effects(state)
    if state.boss_hp <= 0: return Result.WIN
    hit = max(boss_damage - state.player_armor, 1)
    state.player_hp -= hit
    if state.player_hp <= 0: return Result.LOSE

    return Result.CONTINUE

def play(boss_hp, boss_damage, hardmode):
    spells = [MAGIC_DRAIN, MAGIC_SHIELD, MAGIC_POISON, MAGIC_RECHARGE]
    queue = []
    _min = float("inf")

    # --- Search winning games ---
    state = State(PLAYER_HP, PLAYER_MANA, boss_hp)
    queue.append(state)
    while len(queue) > 0:
        state = queue.pop()

        # --- Ignore worst paths ---
        if state.used_mana >= _min: continue

        # --- Hardmode ---
        if hardmode:
            state.player_hp -= 1
            if state.player_hp <= 0: continue

        # --- Apply player turn effects ---
        apply_effects(state)
        if state.boss_hp <= 0:
            _min = min(_min, state.used_mana)
            continue

        # --- Try every spell except Missile ---
        for spell in spells:
            if can_cast(state, spell):
                clone = copy.deepcopy(state)
                #clone = state.clone()
                result = cast(clone, spell, boss_damage)
                if result == Result.WIN: _min = min(_min, clone.used_mana)
                elif result == Result.CONTINUE: queue.append(clone)

        # --- Try cast Missile ---
        if can_cast(state, MAGIC_MISSILE):
            result = cast(state, MAGIC_MISSILE, boss_damage)
            if result == Result.WIN: _min = min(_min, state.used_mana)
            elif result == Result.CONTINUE: queue.append(state)

    return _min

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
text = file.read().replace("\n", " ").strip()
file.close()
cap = re.search(REGEX, text)
boss_hp = int(cap[1])
boss_damage = int(cap[2])

# --- Puzzle 1 ---
_min = play(boss_hp, boss_damage, False)
print(f"1. Least amount of mana: {_min:,d}")

# --- Puzzle 2 ---
_min = play(boss_hp, boss_damage, True)
print(f"2. Least amount of mana: {_min:,d}")
